runner/__init__.py

```diff
@@ -47,107 +75,164 @@ def validate_input_files(workdir):
-        f"--database-url={database_url}",
-        f"--output-dir={output_path}",
-    ]
-
-    os.chdir(workdir)
-    logging.info(f"Running subdocker cmd `{' '.join(cmd)}`")
-    result = subprocess.run(cmd, capture_output=True, encoding="utf8")
-    if result.returncode == 0:
-        log = logging.info
-    else:
-        log = logging.error
-    log(f"cohort-extractor subdocker stdout: {result.stdout}")
-    log(f"cohort-extractor subdocker stderr: {result.stderr}")
-    result.check_returncode()
-    return output_path
+
+
+def run_cohort_extractor(job):
+    joblogger = get_job_logger(job)
+    joblogger.info(f"Starting job")
+    repo = job["repo"]
+    tag = job["tag"]
+    db = job["db"]
+    set_auth()
+    database_url = os.environ[f"{db.upper()}_DATABASE_URL"]
+    with tempfile.TemporaryDirectory(
+        dir=os.environ["OPENSAFELY_RUNNER_STORAGE_BASE"]
+    ) as tmpdir:
+        os.chdir(tmpdir)
+        volume_name = make_volume_name(repo, tag, db)
+        workdir = os.path.join(tmpdir, volume_name)
+        fetch_study_source(workdir, job)
+        validate_input_files(workdir)
+        joblogger.info(f"Repo at {workdir} successfully validated")
+
+        # If running this within a docker container, the storage base
+        # should be a volume mounted from the docker host
+        storage_base = Path(os.environ["OPENSAFELY_RUNNER_STORAGE_BASE"])
+        # We create `output_path` and then map it straight through to the
+        # inner docker container, so that docker-within-docker can write
+        # straight through to the (optionally-mounted) storage base
+        output_path = storage_base / volume_name
+        output_path.mkdir(parents=True, exist_ok=True)
+        # By setting the name to the volume_name, we are guaranteeing only
+        # one identical job can run at once
+        container_name = re.sub(r"[^a-zA-Z0-9]", "-", volume_name)
+        cmd = [
+            "docker",
+            "run",
+            "--name",
+            container_name,
+            "--rm",
+            "--log-driver",
+            "none",
+            "-a",
+            "stdout",
+            "-a",
+            "stderr",
+            "--volume",
+            f"{output_path}:{output_path}",
+            "--volume",
+            f"{workdir}/analysis:/workspace/analysis",
+            "--volume",
+            f"{workdir}/codelists:/workspace/codelists",
+            "docker.pkg.github.com/opensafely/cohort-extractor/cohort-extractor:latest",
+            "generate_cohort",
+            f"--database-url={database_url}",
+            f"--output-dir={output_path}",
+        ]
+
+        os.chdir(workdir)
+        joblogger.info("Running subdocker cmd `%s` in %s", cmd, workdir)
+        result = subprocess.run(cmd, capture_output=True, encoding="utf8")
+        if result.returncode == 0:
+            joblogger.info("cohort-extractor subdocker stdout: %s", result.stdout)
+        else:
+            raise CohortExtractorError(result.stderr)
```

Comment:
Since we include `result.stderr` in the JSON that we `PUT` later on, we need to be really careful here.  Is there a risk that `result.stderr` could contain something sensitive?

Reply:
Yes!

Reply:
We could make `OpenSafelyError` accept two arguments, something like `message` and `sensitive_info` and then have the both appear in the logs, but only the former in the reported error.